aws/s3.go

```diff
@@ -0,0 +1,306 @@
+			logging.Logger.Debugf("Creating session - Bucket %s - region %s", *bucket.Name, bucketRegion)
+			awsSession, err := session.NewSession(&awsgo.Config{
+				Region: awsgo.String(bucketRegion)},
+			)
+			if err != nil {
+				return bucketNamesPerRegion, errors.WithStackTrace(err)
+			}
+			regionClients[bucketRegion] = s3.New(awsSession)
+		}
+
+		validTags, excludeReason := hasValidTags(regionClients[bucketRegion], bucket)
+		if !validTags {
+			logging.Logger.Debugf("Skipping - Bucket %s - %s", *bucket.Name, excludeReason)
+			continue
+		}
+
+		if _, ok := bucketNamesPerRegion[bucketRegion]; !ok {
+			bucketNamesPerRegion[bucketRegion] = []*string{}
+		}
+
+		bucketNamesPerRegion[bucketRegion] = append(bucketNamesPerRegion[bucketRegion], bucket.Name)
+	}
+
+	return bucketNamesPerRegion, nil
+}
+
+func getS3BucketObjects(svc *s3.S3, bucketName *string, isVersioned bool) ([]*s3.ObjectIdentifier, error) {
+	identifiers := []*s3.ObjectIdentifier{}
+
+	// Handle versioned buckets.
+	if isVersioned {
+		err := svc.ListObjectVersionsPages(
+			&s3.ListObjectVersionsInput{
+				Bucket: bucketName,
+			},
+			func(page *s3.ListObjectVersionsOutput, lastPage bool) (shouldContinue bool) {
+				for _, obj := range page.Versions {
+					logging.Logger.Debugf("Bucket %s object %s version %s", *bucketName, *obj.Key, *obj.VersionId)
+					identifiers = append(identifiers, &s3.ObjectIdentifier{
+						Key:       obj.Key,
+						VersionId: obj.VersionId,
+					})
+				}
+				for _, obj := range page.DeleteMarkers {
+					logging.Logger.Debugf("Bucket %s object %s DeleteMarker %s", *bucketName, *obj.Key, *obj.VersionId)
+					identifiers = append(identifiers, &s3.ObjectIdentifier{
+						Key:       obj.Key,
+						VersionId: obj.VersionId,
+					})
+				}
+				return true
+			},
+		)
+		return identifiers, err
+	}
+
+	// Handle non versioned buckets.
+	err := svc.ListObjectsV2Pages(
+		&s3.ListObjectsV2Input{
+			Bucket: bucketName,
+		},
+		func(page *s3.ListObjectsV2Output, lastPage bool) (shouldContinue bool) {
+			for _, obj := range page.Contents {
+				logging.Logger.Debugf("Bucket %s object %s", *bucketName, *obj.Key)
+				identifiers = append(identifiers, &s3.ObjectIdentifier{
+					Key: obj.Key,
+				})
+			}
+			return true
+		},
+	)
+
+	return identifiers, err
+}
+
+func nukeAllS3BucketObjects(svc *s3.S3, bucketName *string, batchSize int) error {
+	versioningResult, err := svc.GetBucketVersioning(&s3.GetBucketVersioningInput{
+		Bucket: bucketName,
+	})
+	if err != nil {
+		return err
+	}
+
+	isVersioned := versioningResult.Status != nil && *versioningResult.Status == "Enabled"
+
+	objects, err := getS3BucketObjects(svc, bucketName, isVersioned)
+	if err != nil {
+		return err
+	}
+
+	totalObjects := len(objects)
+
+	if totalObjects == 0 {
+		logging.Logger.Infof("Bucket: %s - empty - skipping object deletion", *bucketName)
+		return nil
+	}
+
+	if batchSize <= 0 || batchSize > 1000 {
+		batchSize = 1000
+	}
```

Comment:
I think we will want to return an error instead of silently adjusting to the max.

Reply:
Done.